File: src/voyage_geo/stages/analysis/analyzers/mindshare.py
"""Mindshare analyzer — how much of the conversation does the brand own."""

from __future__ import annotations

from collections import Counter

from voyage_geo.types.analysis import MindshareScore
from voyage_geo.types.brand import BrandProfile
from voyage_geo.types.result import QueryResult
from voyage_geo.utils.text import count_occurrences


class MindshareAnalyzer:
    name = "mindshare"

    def analyze(
        self,
        results: list[QueryResult],
        profile: BrandProfile,
        extracted_competitors: list[str] | None = None,
    ) -> MindshareScore:
        valid = [r for r in results if not r.error and r.response]
        if not valid:
            return MindshareScore()

        competitors = extracted_competitors if extracted_competitors else profile.competitors
        all_brands = [profile.name] + competitors
        brand_counts: Counter[str] = Counter()

        for r in valid:
            for brand in all_brands:
                brand_counts[brand] += count_occurrences(r.response, brand)

        total_mentions = sum(brand_counts.values())
        our_mentions = brand_counts.get(profile.name, 0)
        overall = our_mentions / total_mentions if total_mentions > 0 else 0

        # Rank
        sorted_brands = brand_counts.most_common()
        rank = next((i + 1 for i, (b, _) in enumerate(sorted_brands) if b == profile.name), 0)

        # By provider
        by_provider: dict[str, float] = {}
        provider_groups: dict[str, list[QueryResult]] = {}
        for r in valid:
            provider_groups.setdefault(r.provider, []).append(r)
        for prov, prov_results in provider_groups.items():
            prov_total = sum(count_occurrences(r.response, b) for r in prov_results for b in all_brands)
            prov_ours = sum(count_occurrences(r.response, profile.name) for r in prov_results)
            by_provider[prov] = prov_ours / prov_total if prov_total > 0 else 0

        return MindshareScore(
            overall=round(overall, 4),
            by_provider={k: round(v, 4) for k, v in by_provider.items()},
            rank=rank,
            total_brands_detected=len([b for b, c in brand_counts.items() if c > 0]),
        )

File: src/stages/analysis/analyzers/mindshare.ts
import type { Analyzer } from './types.js';
import type { QueryResult } from '../../../types/result.js';
import type { BrandProfile } from '../../../types/brand.js';
import type { MindshareScore } from '../../../types/analysis.js';
import { containsBrand, extractBrandMentions } from '../../../utils/text.js';
import { groupBy, percentage } from '../statistics.js';

export class MindshareAnalyzer implements Analyzer {
  name = 'mindshare';

  analyze(results: QueryResult[], profile: BrandProfile): MindshareScore {
    const validResults = results.filter((r) => !r.error && r.response);
    const allBrands = [profile.name, ...profile.competitors];

    // Count responses mentioning each brand
    const brandMentionCounts = new Map<string, number>();
    for (const brand of allBrands) {
      let count = 0;
      for (const result of validResults) {
        if (containsBrand(result.response, brand)) {
          count++;
        }
      }
      brandMentionCounts.set(brand, count);
    }

    const totalMentions = Array.from(brandMentionCounts.values()).reduce((s, v) => s + v, 0);
    const brandMentions = brandMentionCounts.get(profile.name) || 0;
    const overall = totalMentions > 0 ? percentage(brandMentions, totalMentions) : 0;

    // By provider
    const byProvider: Record<string, number> = {};
    const providerGroups = groupBy(validResults, (r) => r.provider);
    for (const [provider, providerResults] of providerGroups) {
      const mentions = extractBrandMentions(
        providerResults.map((r) => r.response).join(' '),
        allBrands,
      );
      const providerTotal = Array.from(mentions.values()).reduce((s, v) => s + v, 0);
      const providerBrand = mentions.get(profile.name) || 0;
      byProvider[provider] = providerTotal > 0 ? percentage(providerBrand, providerTotal) : 0;
    }

    // By category (using query text to infer category)
    const byCategory: Record<string, number> = {};

    // Rank
    const sorted = Array.from(brandMentionCounts.entries())
      .sort(([, a], [, b]) => b - a);
    const rank = sorted.findIndex(([name]) => name === profile.name) + 1;

    return {
      overall,
      byProvider,
      byCategory,
      rank: rank || 1,
      totalBrandsDetected: allBrands.length,
    };
  }
}
